src/vi/truyenhentai18/src/eu/kanade/tachiyomi/extension/vi/truyenhentai18/TruyenHentai18Dto.kt

```
@@ -0,0 +1,55 @@
+package eu.kanade.tachiyomi.extension.vi.truyenhentai18
+
+import eu.kanade.tachiyomi.source.model.SChapter
+import eu.kanade.tachiyomi.source.model.SManga
+import keiyoushi.utils.tryParse
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import java.text.SimpleDateFormat
+import java.util.Locale
+
+@Serializable
+class SearchDto(
+    val data: List<MangaDto>,
+)
+
+@Serializable
+class MangaDto(
+    val title: String,
+    val slug: String,
+) {
+    fun toSManga() = SManga.create().apply {
+        title = this@MangaDto.title
+        url = "/vi/$slug.html"
```

are old urls compatible?

---

> are old urls compatible?

Yes, example:
Wayback: https://web.archive.org/web/20250311232714/https://truyenhentai18.pro/the-big-breasted-pirate-4.html

old path segment (https://truyenhentai18.app/vi/the-big-breasted-pirate-4.html) redirect to https://truyenhentai18.app/vi/the-big-breasted-pirate-4.html

---

If you silently use the new URL format in the `url` field (which serves as the key of an entry), existing entries might work seamlessly, but when people search for the same title in the source, there will be a new, duplicate entry with the new URL (with the `/vi` prefix). This is not ideal.

Since the manga slug is compatible, you should actually strip the `/vi` prefix when storing the URL and add them back when doing network requests (and in `getMangaUrl()`, etc).

Edit: Same for chapters.

---

> If you silentl ...

The prefix `vi` is added in searchMangaParse, but when popular or latest are used, this prefix defaults to `::popularMangaFromElement`, so perhaps forcing the entry to be updated in mangaDetails could be an alternative to maintain consistency

---

> so perhaps forcing the entry to be updated in mangaDetails could be an alternative to maintain consistency

The app never updates an entry's `url` field, even if you update it in later detail fetches - it is ignored. The `url` functions as an immutable key - different `url` values are recognized as different entries.